Preferences without restart; simpler build process and uncovered a bug with tabbed browsing

matatk opened the pull request:
I've done some cleanup of the build system, UI and have also made the extension reflect changes in preferences as they are changed (i.e. without a browser restart).

However this has shown up a bug that seems quite fundamental to how this (and maybe other) extensions work: when the browser tab changes, we loose the state of the current page (i.e. `previousSelectedIndex`).  If we then go back to that page later, it is possible to have a border "stuck" on an element until we cycle back through it.

I noticed this when implementing preferences being reflected instantly because opening the Add-Ons tab triggers this behaviour.  If one has the Add-Ons tab open always, the border preferences are properly changed in real-time.  If one opens the Add-Ons tab to change the preferences, the state of the web page tab is overwritten, so the code doesn't know which element to remove/add a border from/to when the preference changes.

I am not sure what is the best way to combat this -- keep state for each open tab, or maybe there is a mode for Firefox extensions that runs code specific to each tab (I would hope for the latter).

Perhaps the best way to proceed, if you have time and are able to work on this, would be for me to file that bug on your repo (as that's the root) to go into this further?  If you are not able to work on it I'll just file the bug on mine and try to hack away at it, and make a pull request if/when I fix it.

Please let me know which works for you.

matatk committed:
Removed redundant skin definition.

matatk committed:
Reflect preference changes without browser restart

matatk committed:
Minor options UI tweaks.

matatk committed:
Simplify build.

matatk committed:
Separation of concerns; build script.

* Adding border is now a separate function.
* Code surrounding border prefs cleanup (this has uncovered the
  tab state problem).
* Build script removes Finder hidden files before zipping.
* Version bump in install.rdf.

matatk committed:
Menu label; build robustness; isolate tab bug

* Menu label had got mangled due to Tidy.
* Build script will error out on the first problem.
* Document tab state bug.

davidtodd commented:
Thanks for your work. Please go ahead and file a bug on your repo. It was some time ago that I worked on this and fixing the bug would essentially mean relearning Firefox extension development, which I don't have the cycles for presently.

---

 From: matatk notifications@github.com
To: davidtodd/landmarks landmarks@noreply.github.com 
Sent: Friday, May 24, 2013 10:19 AM
Subject: [landmarks] Preferences without restart; simpler build process and uncovered a bug with tabbed browsing (#3)

I've done some cleanup of the build system, UI and have also made the extension reflect changes in preferences as they are changed (i.e. without a browser restart).
However this has shown up a bug that seems quite fundamental to how this (and maybe other) extensions work: when the browser tab changes, we loose the state of the current page (i.e. previousSelectedIndex).  If we then go back to that page later, it is possible to have a border "stuck" on an element until we cycle back through it.
I noticed this when implementing preferences being reflected instantly because opening the Add-Ons tab triggers this behaviour.  If one has the Add-Ons tab open always, the border preferences are properly changed in real-time.  If one opens the Add-Ons tab to change the preferences, the state of the web page tab is overwritten, so the code doesn't know which element to remove/add a border from/to when the preference changes.
I am not sure what is the best way to combat this -- keep state for each open tab, or maybe there is a mode for Firefox extensions that runs code specific to each tab (I would hope for the latter).
Perhaps the best way to proceed, if you have time and are able to work on this, would be for me to file that bug on your repo (as that's the root) to go into this further?  If you are not able to work on it I'll just file the bug on mine and try to hack away at it, and make a pull request if/when I fix it.
Please let me know which works for you.

---

You can merge this Pull Request by running
git pull https://github.com/matatk/landmarks master
Or view, comment on, or merge it at:
  https://github.com/davidtodd/landmarks/pull/3
Commit Summary
    \* Removed redundant skin definition.
    \* Reflect preference changes without browser restart
    \* Minor options UI tweaks.
    \* Simplify build.
    \* Separation of concerns; build script.
    \* Menu label; build robustness; isolate tab bug
File Changes
    \* M build.sh (134) 
    \* D build_mac.sh (155) 
    \* D config_build.sh (11) 
    \* D content/options.xul (125) 
    \* D content/overlay.xul (19) 
    \* M landmarks.xpi (0) 
    \* D skin/overlay.css (4) 
    \* R src/chrome.manifest (4) 
    \* A src/content/options.xul (112) 
    \* R src/content/overlay.js (125) 
    \* A src/content/overlay.xul (15) 
    \* R src/defaults/preferences/defaults.js (0) 
    \* R src/install.rdf (2) 
    \* R src/locale/en-US/overlay.dtd (0) 
Patch Links:
    \* https://github.com/davidtodd/landmarks/pull/3.patch
    \* https://github.com/davidtodd/landmarks/pull/3.diff

matatk commented:
No worries; thanks for all of your work in making the extension in the first place; the structure of it is very robust and I am sure with these small improvements I've made it will be even more useful to people.

I will work on solving that bug in a separate branch so you're "safe" to merge this pull request if you wish.

Please let me know if you _don't_ want me to make further pull requests in future.  I'll assume that I should continue to do so, at points of code stability.

matatk committed:
Support tab switching; cleaner code encapsulation

* React to switching tabs as with loading pages (still doesn’t store
  current landmark across tabs).
* Change method of encapsulating the code to an anonymous function.
* Increase momentary highlight to 500ms from 250ms.
* Always make the menu; don’t wait to be asked via hovering in the GUI.
* Minor code style consistency cleanups.
* NOTE: some debugging info and alert()s have been left in to help watch
  out for some potential errors for now.

matatk committed:
Use GitHub Releases

* Ignore (having removed) the XPI.
* build script won't fail if the XPI's not there to be deleted in the
  first place.

matatk committed:
README updates

* Explain how to get at the releases page.
* Consistent formatting.

matatk committed:
Better feedback to user

* Remove technical error messages and logging.
* Warn when attempting to navigate when no landmarks were found on the
  page (closes #2).
* Version bump.

matatk committed:
UI cleanup and localisation support

* Add alert() to see if the other branch of getHTMLDocReference() is
called.
* Update preferences XUL to use fewer “style” attributes.
* Remove reference to previously-removed overlay.css from overlay.xul.
* Include a short script to build and install on Macs.
* Ready the preferences window for localisation by using strings.

matatk committed:
Apply upstreram's license

* Include the same license as upstream (in various files too).
* Update README to explain how to build the XPI.

matatk committed:
Code tidying; removeBorder timeout; build process

* Remove redundant check for menu
* Fold makeLandmarks() into makeLandmarksInit() (only called once)
* Rename makeLandmarkMenu() to makeLandmarksMenu() for consistency
* Double removeBorder timeout to 1 second
* Add note to query scope of section header check
* Title case in extension name; friendlier description
* Re-indent consistently with tabs (recommend 4-wide)
* Re-wrap README (pass through Pandoc)
* landmarks -> Landmarks for README title
* Clarify structure of README and border defaults
* Lint overlay.js
* Bump dates in licences
* Incorporate version in build

matatk committed:
Consolidate some ID attrs

matatk committed:
Start conversion to a WebExtension

This is incomplete; have tidied up the code quite a bit and got it
basically working to the extent that it can detect the landmark regions
on a page, as before, and print these out to the console.

matatk committed:
Add note about signed add-ons and WebExtension

matatk committed:
Popup now rudimentarily working

matatk committed:
Structure and Logo

* Move the source code to extension/
* Create a temporary logo
* Use grunt-respimg to convert the logo to appropriately-sized PNGs

matatk committed:
Sort out nesting

matatk committed:
landmarkedElemetns and UI

* Make landmarkedElements an array of objects (easier to read)
* Make the UI be a HTML <ul> (not yet nested)

matatk committed:
Nesting

* Use browser style (currently has no effect in normal Firefox)
* Show the landmarks in the form of a nested <ul>

matatk committed:
Vexation

* Add in temporary bkg colour for logo due to image resize issues
* Add in notes to remove that when possible
* Clean up SVG that ends up in extension/ after Grunting
* Add a note to remove that cleanup task when possible

matatk committed:
Borders are (partly) back!

* Put a border on, and focus, the element selected from the pop-up.

matatk committed:
Commands; README

* Declare commands (not supported fully until Firefox 48)
* Minor tweaks to focusing
* Update README to reflect situation wrt support

matatk committed:
Mostly working on Chrome now too

* Still have warning about JSON key
* Still need to ensure correctly-sized PNGs are in each JSON
* The landmarks are not identified until the tab has been refreshed
  (actually this seems to have cleared up)

matatk committed:
Tidying/Robustness

* Use `var` to declare a variable that wasn't using var before
* Move the decision about what to send when there are no landmarks to a
  more suitable place
* Check for a null response (though it comes through as undefined) when
  no landmarks are found

Note: the odd behaviour with Chrome is because Chrome does not re-inject
scripts when the extension is updated/re-loaded. Not sure what to do
about this at the moment, so leaving it for now.

matatk committed:
Switch to DOM functions; tidying

* Remove some log messages and apparently-redundant JSON key.
* Switch to using DOM functions to construct the landmarks UI tree.
* Workaround bug in Firefox pop-up sizing.
* Make the popup bigger.

Note: It now works fine in Firefox Developer Edition, including the
command bindings (not yet functional). However, the popup bug is still
present.

matatk committed:
Build separate versions for Firefox and Chrome

matatk committed:
README

matatk committed:
DRY-up the Gruntfile a bit

* Declare the most repetitive tasks in a loop.
  + JSON merging
  + File copying
  + Building for each browser

Notes:
 - There are still some fairly repetitive tasks.
 - Templating for manifest.json needs to be done.

matatk committed:
Re-order fields in manifest slightly

matatk committed:
Use correct minimum Firefox version

matatk committed:
Improved handling of landmarks sent from content

matatk committed:
Refreshing landmarks; message-passing; styles

* Attempt to get landmarks only when asked.
* Report errors in the popup.
* Firefox Developer Edition suggested setting browser_style in
  manifest.json, so it has been :-).

matatk committed:
Only get landmarks once per tab-URL

matatk committed:
Starting options page; quotes; whitespace

* Add options page (non-functional so far)
* Change double to single quotes in content.js
* Minor whitespace tweak in content.js
* Change double to single quotes in popup.js

matatk committed:
Options

* Add JS to options page.
* Include wrapper to support Firefox.

matatk committed:
Keyboard shortcuts; Equality tests

* Implement keyboard shortcuts
* Use more robust equality tests

matatk committed:
Documentation

* Shortcut keys
* Changing shortcut keys
* Link to issue on content script re-injection from README
* Explain that the extension is not presently available in-browser

matatk committed:
Support single-page apps

* Monitor for use of History API as a heuristic for the page having
  made a significant transition without a reload.
* Partially addresses #16.

matatk committed:
Implement options

matatk committed:
Tidying; ES6

* Switch to switch for the big if
* Use const and Object.freeze()
* Include JSHint RC file
* Use let for global vars (partial progress towards #14)
* Rename globals to be clear about them

matatk committed:
Cleaning; Linting

* Make the clean task browser-specific
* Run the browser-specific clean task before building for that browser
* Use JSHint on the built code

matatk committed:
SVG to PNG

* Create a custom task to convert via ImageMagick; fixes #11
* Update deps accordingly
* Update manifest.json files accordingly
* Remove hacky 'todo' target for clean as it's redundant

matatk committed:
Update SVG

Need to figure out how to do transparency.

matatk committed:
...that was surprisingly easy for once!

matatk committed:
Building zip files; Docs; DRY; Version; Settings

* Build zip files for the extensions; the zip files are output
  to build/<browser>/
* Rename src/build/ to src/assemble/ to better reflect its
  purpose and so as not to be confused with build/
* Use package.json variables and grunt-text-replace to DRY
  version and other variables; fixes #13
* .gitignore the build/ directory
* Depend on the grunt-zip plugin
* Update README
* Bump version to 1.0.0 as it was already stable before WebExtension
* Update description in manifest (and on GitHub site)
* Update homepage URL to the GitHub pages site
* Ensure that momentary is the default value when restoring settings

matatk committed:
Hooks; Ignores; Docs

* Ignore generated site from gh-pages
* Manage the git hook in the repo
* Tidy up the hook a bit from that on SO (link included in the hook)
* Big re-org of README for improved clarity

matatk committed:
Change extension long name for clarity (I hope)

matatk committed:
Disable browser action when not on an HTTP(S) page

This goes some way to addressing #16.

matatk committed:
Re-structure to allow badging

* Content script sends number of landmarks for the background script to
  put on a badge, when it is ready (closes #21).
* Handle URL/navigation events to decide if the browser action should
  be disabled (or not).
* Cleaner structure for handling refreshes.
* Remove spaces before open parentheses after ‘function’.
* Remove quotes around dict key names.

Known limitations:
* In Chrome, if the browser opens on a non-HTTP(S) tab, then the browser
  action is *not* disabled, but should be.
* In Firefox (Dev Ed), the about: tabs do not have their browser action
  disabled. Also observed one instance of a normal page tab being
  disconnected from the content script, as if the script weren't there.

matatk committed:
Clean up content.js

* Consistent function-call spacing
* Revert error message in delta to being appropriately verbose/clear
* Add an alert() for when keyboard shortcuts are used but landmarks
  have not yet been found

matatk committed:
Badging; Messaging; Code and Comment Tidying

* Remove redundant tabs query from background.js
* Move badging code to make reading order clearer in background.js
* Improve comments’ clarity
* Put some bounds on document-ready checking and add some informative
  logging (if ever needed) in content.js
* Content script sends ‘wait’ back to popup.js if the landmarks have
  not been found yet
* Popup.js responds correctly to the wait message.
* Content.js always calls filterLandmarks() even if 0 were found (was a
  bit of a premature optimisation).

matatk committed:
README

* Add more info, more clearly linked, for authors/developers.
* Simplify the first para.
* Explain the Firefox situation a bit more clearly.

matatk committed:
Harmonise copyright message in LICENSE with content.js

matatk committed:
Support for Internationalisation (part 1)

* Put in the infrastructure for i18n, closes #19.
* I have not put in translation strings for the messages that go to
  the console, via console.log() or throw(). The idea is to reduce the
  effort required to make a user-facing translation.
* I have tried to name the messages clearly enough such that the
  "description" field is not necessary. I think this has been achieved,
  but it can be reviewed later, particularly if more strings are added.

Note: prefs and options dialog translation hooks are not yet there.

matatk committed:
Support for Internationalisation (part 2); const/let

* Translate options.html (part-addresses #19)
* Space out the different secitons of message.json
* Move popup script call to the end of the page body
* Move away from using var in options.js (part-addresses #14)
* Move away from using var in popup.js (part-addresses #14)
* Move away from using var in content.js (part-addresses #14)
* Move away from using var in Gruntfile.js (closes #14)

Interesting: there are no variable declarations in background.js!

For now, the user-visible landmark role names are not translated;
they probably should be.

matatk committed:
Switch to using my grunt plugin for SVG conversion

matatk committed:
Copyright and Acknowledgements

* Reflect latest copyright situation.
* Copyright declarations conventionally have the year before the holder.
* Add acknowledgements to README.

matatk committed:
README; Version

* Update README now Firefox 48 is out.
* Put Test Pages as a subsection of Development.
* Add a Changes section.
* Make version 2.0.0 for release, as this is the first WebExtension
  release, and it has been extensively modified, this seems apt.

matatk committed:
Reload page button

* Reload page button closes #16 for now
* Use latest time-grunt
* Clarify comments on watching tabs for URL changes

matatk committed:
Add Firefox screengrab and caption

matatk committed:
Ensure locales are included in ZIP

matatk committed:
More Meta stuff and a version bump

* Add meta stuff for Google
* Had to bump version due to fumbling about in the UIs for Mozilla

matatk committed:
Meta for Chrome

* Add a script to resize the image by putting the screengrab on a
  transparent background; thanks to:
  http://imagemagick.org/discourse-server/viewtopic.php?t=24961
* Git-ignore the generated image.

matatk committed:
Merge branch 'master' into convert-to-webextension

matatk committed:
Merge pull request #25 from matatk/convert-to-webextension

Convert to webextension

matatk committed:
Installation info; metadata

* Add the basic detailed description used for the add-ons pages.
* Update README with installation page links.

matatk committed:
Firefox URL update now review is done

matatk committed:
post-commit hook

* Skip the first three lines of README so that we don't end up with more
  than one <h1> on the page.

matatk committed:
Make sure ACKs are clearly visible in listings

matatk commented:
I realised this is still open, and things have diverged quite a lot now (in the intervening time I've learnt that it's a good idea to make PRs from separate feature branches :-)). I have now got it all, at long last, converted to being a WebExtension (works on Chrome, too) and it's available directly in-browser via the browsers' respective add-on facilities.

Your contribution is, of course, still acknowledged, and I'm grateful to have had your great work to stand on the shoulders of.  I'm hoping that this new version of the extension will be easy for many more people to discover and install in-browser. Thanks!
